BlogPost: Job's Backoff Limit Per Index Goes GA (#49982)
kubernetes/website: content/en/blog/_posts/2025-04-23-jobs-backofflimitperindex-goes-ga.md

---
layout: blog
title: "Kubernetes 1.33: Job's Backoff Limit Per Index Goes GA"
date: 2025-04-23
slug: kubernetes-1-33-jobs-backoff-limit-per-index-goes-ga
author: >
  [Michał Woźniak](https://github.com/mimowo) (Google)
---

This post describes the Job's _Backoff Limit Per Index_, which graduates to
stable in Kubernetes 1.33, and how to use it in your Jobs.

## About

When you run workloads on Kubernetes, you need to account for Pod failures due
to variety of reasons. Ideally, your workload can tolerate some transient
failures and continue running to completion.

One way to achieve failure tolerance in Kubenetes Job is by setting the
`spec.backoffLimit` field which specifies the total number of tolerated
failures.

> However, for Indexed Jobs this is often not flexible enough for embarassingly

Author: Done

parallel workloads, where every index is considered independent, for example

> in case of running integration tests. In that scenario, a fast-failing index is

Author: Yeah, I'm not attached to the initial version. Let me commit yours, and we see if others can come up with adjustments.

likely to consume your entire budget for tolerating Pod failures, and you may
not be able to run the other indexes.

In order to address this limitation we introduce the new feature, called
_Backoff Limit Per Index_, which allows you to control the number of retries
per index.

## How it works

You can specify the number of tolerated Pod failures per index with the new
`spec.backoffLimitPerIndex` field.

When the number of tolerated failures is exceeded the Job considers the index as
failed, and this index is listed in the Job's `status.failedIndexes` field.

Additionally, you can configure short-circuit for detecting a failed index by

> using the "FailIndex" action in the _Pod Failure Policy_ feature (see the

Reviewer: Link to doc page about the feature.

example below).

When the `spec.backoffLimitPerIndex` the Job executes all indexes, unless the cap
of the total number of failed indexes is specified by the `spec.maxFailedIndexes`

> field.

Reviewer: ```
When the `spec.backoffLimitPerIndex` is specified the Job will execute all indexes, 
unless the cap of the total number of failed indexes is specified by the 
`spec.maxFailedIndexes` field.

### Example

The Job spec snippet below demonstrates an example use, combining the feature
with the _Pod Failure Policy_ feature:

```yaml
completions: 10
parallelism: 10
completionMode: Indexed
backoffLimitPerIndex: 1
maxFailedIndexes: 5
podFailurePolicy:
  rules:
  - action: Ignore
    onPodConditions:
    - type: DisruptionTarget
  - action: FailIndex
    onExitCodes:
      operator: In
      values: [ 42 ]
```

In this example, the Job handles Pod failures as follows:

- Ignores any failed Pods that have the built-in `DisruptionTarget`
  condition. These Pods don't count towards backoff limits.
- Fails the index corresponding to the failed Pod if any of the failed Pod's
  containers exited with the exit code 42 - based on the matching "FailIndex"
  rule.

> - Retries the first failure of any index, unless the index if failed due to the

Reviewer: ```suggestion
- Retries the first failure of any index, unless the index failed due to the
```

---
layout: blog
title: "Kubernetes 1.33: Job's Backoff Limit Per Index Goes GA"

> date: 2025-04-23

Reviewer 2: While this blog is about the v1.33 release, we will limit to one blog published per day. This means it is likely that this blog will be planned on a later date, and the schedule is yet to be fully decided. (It is highly dependent on other blog reviews completed in time)

If you can simply add `draft: true` in the front matter, it would allow us to merge the blog without actually publishing them to the website. Once the blog content is finalised and merged, the Release Comms team can take over the rest of the publication setup.

```suggestion
date: 2025-04-23
draft: true
```

You can find more in the blog guideline [here](https://kubernetes.io/docs/contribute/blog/article-submission/#article-scheduling)! 👍

Author: Got it, thanks!

slug: kubernetes-1-33-jobs-backoff-limit-per-index-goes-ga
author: >
  [Michał Woźniak](https://github.com/mimowo) (Google)
---

In Kubernetes v1.33, the _Backoff Limit Per Index_ feature reaches general
availability (GA). This blog describes the Backoff Limit Per Index feature and
its benefits.

## About Backoff Limit Per Index

When you run workloads on Kubernetes, you must consider scenarios where Pod
failures can affect the completion of your workloads. Ideally, your workload
should tolerate transient failures and continue running.

> To achieve failure tolerance in Kubernetes Job, you can set the

Reviewer 3: ```suggestion
To achieve failure tolerance in a Kubernetes Job, you can set the
```

`spec.backoffLimit` field. This field specifies the total number of tolerated
failures.

However, for workloads where every index is considered independent, like
[embarassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)
workloads, the `spec.backoffLimit` field is often not flexible enough if you use

> Indexed Jobs. For example, integration tests. In that scenario, a fast-failing

Reviewer 3: Before you go into the details of the problem, can you explain what an index/indexed job is in this context? Some short explanation and maybe some links would be great

Author: Thanks, I have rephrased the section to make it a bit more clear, and added a link to the relevant guide on using Indexed Jobs for parallel workloads with static work assignments. 

Let me know what you think, I'm happy to adjust it more.

Reviewer 3: LGTM, thanks!

index is likely to consume your entire budget for tolerating Pod failures, and
you might not be able to run the other indexes.

In order to address this limitation, we introduce _Backoff Limit Per Index_,
which allows you to control the number of retries per index.

## How Backoff Limit Per Index works

To use Backoff Limit Per Index for Indexed Jobs, configure the following:
* Specify the number of tolerated Pod failures per index with the
  `spec.backoffLimitPerIndex` field. When you set this field, the Job executes
  all indexes by default.
* Specify the cap on the total number of failed indexes by setting the
  `spec.maxFailedIndexes` field. When the limit is exceeded the entire Job is
  terminated.

> * Short-circuit to detect a failed index by using the `FailIndex` action in the
>   [Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
>   feature.

Reviewer 4: ```suggestion
* Configure a short circuit to detect a failed index by using the `FailIndex` action in the
  [Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
  feature.
```

Reviewer 4: This is just a nit if it makes sense to you.

Reviewer 4: I just wanted to make the list of bullets parallel by using a verb at the begining: specify, specify,...
Maybe 'Define a short circuit to detect a failed index by using the `FailIndex` action in the'?

Again, this is such a nit, you can disregard this thread altogether 😄

Author: I see the point as we start the previous bullet point with Specify. Makes sense. I like Define. I will update it shortly. Thank you!

Author: Done

When the number of tolerated failures is exceeded, the Job marks that index as
failed and lists it in the Job's `status.failedIndexes` field.

### Example

The following Job spec snippet is an example of how to combine Backoff Limit Per
Index with the _Pod Failure Policy_ feature:

```yaml
completions: 10
parallelism: 10
completionMode: Indexed
backoffLimitPerIndex: 1
maxFailedIndexes: 5
podFailurePolicy:
  rules:
  - action: Ignore
    onPodConditions:
    - type: DisruptionTarget
  - action: FailIndex
    onExitCodes:
      operator: In
      values: [ 42 ]
```

In this example, the Job handles Pod failures as follows:

- Ignores any failed Pods that have the built-in `DisruptionTarget`

>   condition. These Pods don't count towards backoff limits.

Reviewer 3: I think it wouldn't hurt to mention a little more of when can this condition occur. But as you prefer, really.

Author: Slightly tweaked and added a link. PTAL.

Reviewer 3: SGTM, thanks!

---
layout: blog
title: "Kubernetes 1.33: Job's Backoff Limit Per Index Goes GA"
date: 2025-04-23
slug: kubernetes-1-33-jobs-backoff-limit-per-index-goes-ga
author: >
  [Michał Woźniak](https://github.com/mimowo) (Google)
---

This post describes the Job's _Backoff Limit Per Index_ feature, which graduates to

> stable in Kubernetes 1.33, and how you can benefit from using it.

Author: Done

## About

Running your workloads on Kubernetes, you need to account for Pod failures due
to variety of reasons. Ideally, your workload should tolerate some transient
failures and continue running to completion.

One way to achieve failure tolerance in Kubenetes Job is by setting the
`spec.backoffLimit` field which specifies the total number of tolerated

> failures.

Author: Done

However, for Indexed Jobs this is often not flexible enough for
[embarassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)

> workloads, where every index is considered independent. For example, integration

Author: Sure, done

tests. In that scenario, a fast-failing index is likely to consume your entire

> budget for tolerating Pod failures, and you may not be able to run the other

Author: Done

indexes.

In order to address this limitation we introduce the new feature, called
_Backoff Limit Per Index_, which allows you to control the number of retries

> per index.

Author: Done

## How it works

> 

Reviewer 4: Suggestion for this section:

```
To use Backoff Limit Per Index, you can configure the following:

* For Indexed Jobs, specify the number of tolerated Pod failures per index by adding the
`spec.backoffLimitPerIndex` field. This field is available only in Indexed Jobs.
* Short circuit to detect a failed index by
using the "FailIndex" action in the
[Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
feature (see the example below).

When the `spec.backoffLimitPerIndex` is specified the Job will execute all
indexes, unless the cap of the total number of failed indexes is specified by
the `spec.maxFailedIndexes` field.

When the number of tolerated failures is exceeded, the Job considers the index as
`failed`. Additionally, this index is listed in the Job's `status.failedIndexes` field.
```

Author: I have mixed the proposal with the current phrasing. One thing I wanted to improve about the proposal is that the ``spec.maxFailedIndexes` field is configurable option which is not mentioned in the bullet points.

You can specify the number of tolerated Pod failures per index with the new
`spec.backoffLimitPerIndex` field. Important to note is that this field will only apply to Indexed Job.

When the number of tolerated failures is exceeded the Job considers the index as
failed, and this index is listed in the Job's `status.failedIndexes` field.

Additionally, you can configure short-circuit for detecting a failed index by
using the "FailIndex" action in the
[Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
feature (see the example below).

When the `spec.backoffLimitPerIndex` is specified the Job will execute all
indexes, unless the cap of the total number of failed indexes is specified by

> the `spec.maxFailedIndexes` field.

Reviewer 5: Do we also want to specify that `.spec.backoffLimit` still applies (or its default) or if it should be used in tandem with the `spec.backoffLimitPerIndex`? Otherwise, the blog lgtm.

Author: Yeah, good question. I was thinking about mentioning this but concluded to myself it is a technical detail, a user can figure out when playing with the feature. 

During the KEP discussion we didn't see a use case for combining backoffLimit with backoffLimitPerIndex, but it is supported. 

So, when, a user sets backoffLimitPerIndex, then API server defaults backoffLimit to MaxInt32. We do this to handle them independently, and yet make sure backoffLimit is not nil, because maybe this would break some clients or 3rd party users of Job which assume backoffLimit is always non-nil.

Reviewer 5: Ok, good to see that the default is reasonable and the user does not have to think about it.

In that case, I guess we do not have to include the override in the blog if it is a more obscure feature. And if the user sets that, they can probably expect the behavior.

Nevertheless, it might make sense to add it to the docs.

Author: The defaulting code is here: https://github.com/kubernetes/kubernetes/blob/908bdb3f2c487061975b9a06ca0480e3388bc285/pkg/apis/batch/v1/defaults.go#L44-L50

The reason we considered it a technical detail is that the use-case could be to terminate Job execution early if there are many failures, but assuming the indexes are independent `spec.maxFailedIndexes` can solve this use-case.

> Nevertheless, it might make sense to add it to the docs.

I can see two places to update:
- API comment for the [backoffLimit field](https://github.com/kubernetes/kubernetes/blob/908bdb3f2c487061975b9a06ca0480e3388bc285/pkg/apis/batch/types.go#L351) as we say "Defaults to 6", plus probably also mention on the backoffLimitPerIndex field.
- documentation in https://kubernetes.io/docs/concepts/workloads/controllers/job/, as we say "The back-off limit is set by default to 6.".

I will open PRs, but it is too late for 1.33 I suppose. Let me know if you have more suggestions which places to update.

Reviewer 5: Sounds good, thanks!

Author: Opened the PRs:
- https://github.com/kubernetes/kubernetes/pull/131256
- https://github.com/kubernetes/website/pull/50461

PTAL

### Example

> The Job spec snippet below demonstrates an example use, combining the feature
> with the _Pod Failure Policy_ feature:

Reviewer 4: ```suggestion
The following Job spec snippet is an example of how to combine Backoff Limit Per Index with the _Pod Failure Policy_ feature:
```

Author: done

```yaml
completions: 10
parallelism: 10
completionMode: Indexed
backoffLimitPerIndex: 1
maxFailedIndexes: 5
podFailurePolicy:
  rules:
  - action: Ignore
    onPodConditions:
    - type: DisruptionTarget
  - action: FailIndex
    onExitCodes:
      operator: In
      values: [ 42 ]
```

In this example, the Job handles Pod failures as follows:

- Ignores any failed Pods that have the built-in `DisruptionTarget`
  condition. These Pods don't count towards backoff limits.
- Fails the index corresponding to the failed Pod if any of the failed Pod's
  containers finished with the exit code 42 - based on the matching "FailIndex"
  rule.
- Retries the first failure of any index, unless the index failed due to the

>   matching FailIndex rule.

Reviewer 4: ```suggestion
  matching `FailIndex` rule.
```

Author: done

---
layout: blog

> title: "Kubernetes 1.33: Job's Backoff Limit Per Index Goes GA"

Reviewer 6: Probably a nit (fine to leave as-is since it’s a draft), but worth aligning  to our standard release title format.

```suggestion
title: "Kubernetes v1.33: Job's Backoff Limit Per Index Goes GA"
```

date: 2025-04-23
draft: true

> slug: kubernetes-1-33-jobs-backoff-limit-per-index-goes-ga

Reviewer 6: ```suggestion
slug: kubernetes-v1-33-jobs-backoff-limit-per-index-goes-ga
```

author: >
  [Michał Woźniak](https://github.com/mimowo) (Google)
---

In Kubernetes v1.33, the _Backoff Limit Per Index_ feature reaches general
availability (GA). This blog describes the Backoff Limit Per Index feature and
its benefits.

## About Backoff Limit Per Index

When you run workloads on Kubernetes, you must consider scenarios where Pod
failures can affect the completion of your workloads. Ideally, your workload
should tolerate transient failures and continue running.

To achieve failure tolerance in a Kubernetes Job, you can set the
`spec.backoffLimit` field. This field specifies the total number of tolerated
failures.

However, for workloads where every index is considered independent, like
[embarassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)
workloads - the `spec.backoffLimit` field is often not flexible enough.
For example, you may choose to run multiple suites of integration tests by

> representing each suite as an index within an [Indexed Job](https://kubernetes.io/docs/tasks/job/indexed-parallel-processing-static/).

Reviewer 6: ```suggestion
representing each suite as an index within an [Indexed Job](/docs/tasks/job/indexed-parallel-processing-static/).
```

In that setup, a fast-failing index  (test suite) is likely to consume your
entire budget for tolerating Pod failures, and you might not be able to run the
other indexes.

In order to address this limitation, we introduce _Backoff Limit Per Index_,
which allows you to control the number of retries per index.

## How Backoff Limit Per Index works

To use Backoff Limit Per Index for Indexed Jobs, configure the following:

> * Specify the number of tolerated Pod failures per index with the

Reviewer 4: I'd suggest this to be a numbered list instead of a bulleted list, because these are sequential steps (I assume)

Author: Actually, they don't have a clear order, they are 3 things a user may want to configure. Settings 2 and 3 are optional for using the feature.

 wdyt?

Reviewer 4: So maybe, I'd suggest something similar to this:

```suggestion
To use Backoff Limit Per Index for Indexed Jobs, configure specify the number of tolerated Pod failures per index with the  `spec.backoffLimitPerIndex` field. When you set this field, the Job executes  all indexes by default.
If your use case requires additional limit and fallback configurations, you can also complete the following tasks:
* Specify the cap on the total number of failed indexes by setting the
  `spec.maxFailedIndexes` field. When the limit is exceeded the entire Job is
  terminated.
* Define a short-circuit to detect a failed index by using the `FailIndex` action in the
  [Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
  feature.
```

Author: looks better, just a small tweaks I would suggest maybe:

```
To use Backoff Limit Per Index for Indexed Jobs, specify the number of tolerated
Pod failures per index with the `spec.backoffLimitPerIndex` field. When you set
this field, the Job executes all indexes by default.

Additionally, to fine-tune the error handling:
* Specify the cap on the total number of failed indexes by setting the
  `spec.maxFailedIndexes` field. When the limit is exceeded the entire Job is
  terminated.
* Define a short-circuit to detect a failed index by using the `FailIndex` action in the
  [Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
  feature.
```
So that:
1. avoid `configure specify`
2. avoid saying `fallback configurations` because this is not necessarily the intention of the FailIndex action

wdyt?

Reviewer 4: LGTM!

---
layout: blog
title: "Kubernetes v1.33: Job's Backoff Limit Per Index Goes GA"
date: 2025-04-23
draft: true
slug: kubernetes-v1-33-jobs-backoff-limit-per-index-goes-ga
author: >
  [Michał Woźniak](https://github.com/mimowo) (Google)
---

In Kubernetes v1.33, the _Backoff Limit Per Index_ feature reaches general
availability (GA). This blog describes the Backoff Limit Per Index feature and
its benefits.

> ## About Backoff Limit Per Index

Reviewer 7: nit:
```suggestion
## About backoff limit per index
```

When you run workloads on Kubernetes, you must consider scenarios where Pod
failures can affect the completion of your workloads. Ideally, your workload
should tolerate transient failures and continue running.

To achieve failure tolerance in a Kubernetes Job, you can set the
`spec.backoffLimit` field. This field specifies the total number of tolerated
failures.

However, for workloads where every index is considered independent, like
[embarassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)
workloads - the `spec.backoffLimit` field is often not flexible enough.
For example, you may choose to run multiple suites of integration tests by
representing each suite as an index within an [Indexed Job](/docs/tasks/job/indexed-parallel-processing-static/).
In that setup, a fast-failing index  (test suite) is likely to consume your
entire budget for tolerating Pod failures, and you might not be able to run the
other indexes.

> In order to address this limitation, we introduce _Backoff Limit Per Index_,

Reviewer 7: ```suggestiom
In order to address this limitation, Kubernetes introduced _backoff limit per index_,
```

which allows you to control the number of retries per index.

> ## How Backoff Limit Per Index works

Reviewer 7: ```suggestion
## How backoff limit per index works
```

To use Backoff Limit Per Index for Indexed Jobs, specify the number of tolerated
Pod failures per index with the `spec.backoffLimitPerIndex` field. When you set
this field, the Job executes all indexes by default.

Additionally, to fine-tune the error handling:
* Specify the cap on the total number of failed indexes by setting the
  `spec.maxFailedIndexes` field. When the limit is exceeded the entire Job is
  terminated.
* Define a short-circuit to detect a failed index by using the `FailIndex` action in the
  [Pod Failure Policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)

>   feature.

Reviewer 7: (nit)
```suggestion
  mechanism.
```

When the number of tolerated failures is exceeded, the Job marks that index as
failed and lists it in the Job's `status.failedIndexes` field.

### Example

The following Job spec snippet is an example of how to combine Backoff Limit Per

> Index with the _Pod Failure Policy_ feature:

Reviewer 7: nit: we use sentence case (don't write feature names in title case)

Rationale for this is that lots of things arrived as features (eg: tolerations; indices for Job Pods; sidecars; init containers; some firms of garbage collection). We don't write features in title case because there are so many of them.

  containers finished with the exit code 42 - based on the matching "FailIndex"
  rule.
- Retries the first failure of any index, unless the index failed due to the
  matching `FailIndex` rule.
- Fails the entire Job if the number of failed indexes exceeded 5 (set by the
  `spec.maxFailedIndexes` field).

## Learn more

- Read the blog post on the closely related feature of Pod Failure Policy [Kubernetes 1.31: Pod Failure Policy for Jobs Goes GA](/blog/2024/08/19/kubernetes-1-31-pod-failure-policy-for-jobs-goes-ga/)
- For a hands-on guide to using Pod failure policy, including the use of FailIndex, see
  [Handling retriable and non-retriable pod failures with Pod failure policy](/docs/tasks/job/pod-failure-policy/)
- Read the documentation for
  [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index) and
  [Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
- Read the KEP for the [Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs)

## Get involved

> This work was sponsored by

Reviewer 7: ```suggestion
This work was sponsored by the Kubernetes
```
?
